src/controller/stream.rs

```diff
@@ -0,0 +1,358 @@
+}
+
+struct PartialResult {
+    data: Vec<u8>,
+    next_range: Range,
+}
+
+impl StreamController {
+    pub fn new(
+        request: ClientRequest,
+        stream_id: usize,
+        network: Arc<NetworkClient>,
+    ) -> Result<Self, RequestError> {
+        let first_chunk = network.find_chunk(&request.dataset_id, request.query.first_block());
+        let Some(first_chunk) = first_chunk else {
+            tracing::info!(
+                "No chunk found for block {}, dataset {}",
+                request.query.first_block(),
+                request.dataset_id
+            );
+            return Err(RequestError::NotFound(format!(
+                "No chunk found for block {}",
+                request.query.first_block()
+            )));
+        };
+
+        let buffer = SlidingArray::with_capacity(request.buffer_size);
+        let timeout_quantile = request.timeout_quantile;
+        Ok(Self {
+            request,
+            stream_id,
+            network,
+            next_index: 0,
+            next_chunk: Some(first_chunk),
+            downloads: Default::default(),
+            buffer,
+            timeouts: Arc::new(TimeoutManager::new(timeout_quantile)),
+        })
+    }
+
+    #[instrument(skip_all, fields(stream_id = self.stream_id))]
+    pub async fn poll_next_chunk(&mut self) -> Option<Result<ResponseChunk, RequestError>> {
+        while self.buffer.len() < self.request.buffer_size && self.next_chunk.is_some() {
+            self.schedule_next_chunk();
+        }
+
+        while !self.buffer.front().is_some_and(Slot::ready) {
+            let (result, index) = self
+                .downloads
+                .join_next()
+                .await?
+                .expect("Awaiting download panicked");
+            self.buffer[index] = result;
+
+            tracing::debug!(
+                "Got result for chunk number {index} ({}), buffer:\n[{}]",
+                short_code(&self.buffer[index]),
+                self.buffer
+                    .data()
+                    .iter()
+                    .map(Slot::debug_symbol)
+                    .collect::<String>()
+            );
+
+            if let Slot::Done(Err(_)) = self.buffer[index] {
+                // The stream will end at the first error. There is no need to start new downloads.
+                self.next_chunk = None;
+            }
+        }
+
+        let index = self.buffer.first_index();
+        let result = match self.buffer.pop_front()? {
+            Slot::Done(result) => Some(result),
+            Slot::Partial(PartialResult { data, next_range }) => {
+                self.buffer.push_front(Slot::Pending);
+                let future = self
+                    .get_chunk_response(index, next_range)
+                    .map(move |result| (result, index));
+                self.downloads.spawn(future);
+                Some(Ok(data))
+            }
+            Slot::Pending => unreachable!("First chunk is not ready"),
+        };
```

Comment:
We popped the chunk, but haven't scheduled a new download. As a result we are downloading with less concurrency than requested.